source/ncengine/graphics2/diligent/pass/PassBackend.cpp

```diff
@@ -174,23 +182,39 @@ void PassBackend::Update(const PostProcessState& postProcessState)
 void PassBackend::RenderMaterial(Diligent::IDeviceContext& context,
                                  Diligent::ISwapChain& swapChain,
                                  PerPassResourceSignature& perPassResourceSignature,
-                                 const std::vector<std::vector<Batch>>& passBatches)
+                                 const std::vector<std::vector<Batch>>& staticPassBatches,
+                                 const std::vector<std::vector<Batch>>& skinnedPassBatches)
 {
     NC_PROFILE_SCOPE("PassBackend::RenderMaterial()", ProfileCategory::Rendering);
-    NC_ASSERT(m_materialPasses.size() == passBatches.size(), "Frontend/Backend passes out of sync.");
+    NC_ASSERT(
+        m_staticMaterialPasses.size() == staticPassBatches.size() &&
+        m_skinnedMaterialPasses.size() == skinnedPassBatches.size(),
+        "Frontend/Backend passes out of sync."
+    );
+
+    auto passView = std::views::zip(
+        m_staticMaterialPasses,
+        m_skinnedMaterialPasses,
+        staticPassBatches,
+        skinnedPassBatches
+    );
 
-    for (auto [pass, batches] : std::views::zip(m_materialPasses, passBatches))
+    for (auto [staticPass, skinnedPass, staticBatches, skinnedBatches] : passView)
     {
+        // PassManifest verifies static/skinned pass pairs specify the same render targets, so we can just choose from either here.
         BindRenderTarget(context, swapChain,
                          perPassResourceSignature.GetPostProcessColorSinkBufferResource(),
                          perPassResourceSignature.GetPostProcessDepthSinkBufferResource(),
-                         pass.colorRTIndex, pass.depthRTIndex);
+                         staticPass.colorRTIndex, staticPass.depthRTIndex);
         ClearRenderTarget(context, swapChain,
```

Comment:
Quick Q: Are material passes pretty much guaranteed to always use different render targets? Was wondering if this clear could potentially wipe out a previous pass.

Reply:
This is still a layer of fragility I don't have a great solution for. You've found the fragility, though are backwards on why it's a problem:

Material passes were designed to always use the **_same_** render targets. I didn't know how to express this in the code, but any contiguous (the pass descs are contiguous in the manifest) passes that share the same sinks will be merged in the API into the same `vkRenderPass`. 

All passes considered a `MaterialPass` have the backends call the 'ClearRenderTarget' instruction because they get merged into a single pass, and this pass clears the `MainColor` render target before drawing, which effectively wipes the last frame's pixels.
 
The "Misc Passes" don't clear the render target at all, so they're intended to be 'additive' passes that overlay on top of   an existing pass.

The post process passes all clear their render targets, because they're expected to each have their own pass. 

I would love a sit down with you to talk about a better way to enforce this in code.

Here are the underlying behaviors/invariants:

- All passes considered a `MaterialPass` (currently skinned and static renderers) with the same Sink must be ordered contiguously, because the contiguous block of passes will be merged, and the merged pass will clear the Sink target. A material pass can be defined to output to a separate sink, it will clear that sink before rendering to it and is given its own `vkRenderPass`.
- All passes considered a 'MiscPass' do not clear their Sink target.
- All passes considered a 'PostProcessPass' clear their Sink target. They must have separate sinks, or they will overwrite the previous sink.

This is as robust as I could make it without a proper render graph: the hardcoded final pass always has a new render target, and it clears this before reading from (if any) the sources output by any previous pass (as a separate render target shader input), so if we have no material passes the post process passes all clear themselves so we don't get any compositing.

Thinking about this now - if we have no material passes but did have a Misc Pass, the Misc Passes _would_ continually add on to themselves so we'd get the output of the previous frames endlessly added on. I'll add an issue for Misc Passes to check and clear the target if there are no material passes.

![Passes](https://github.com/user-attachments/assets/842a91f5-3b07-45d0-ab47-c484e99f2a82)